File: Sources/CS2TextureCombiner/Services/TexturePacking.swift
import Foundation

enum TexturePacking {
    static func validateBaseColor(
        _ input: InputMap?,
        assetType: AssetType = .building
    ) throws -> PixelSize {
        guard let input else { throw CombinerError.baseColorRequired }
        try validateMainInputSize(input.size, name: input.slot.title, assetType: assetType)
        return input.size
    }

    static func validateMainInputSize(
        _ size: PixelSize,
        name: String,
        assetType: AssetType = .building
    ) throws {
        guard size.isSquare, assetType.allowedSizes.contains(size.width) else {
            if assetType != .building {
                throw CombinerError.invalidProfileTextureSize(
                    profile: assetType.rawValue,
                    name: name,
                    size: size,
                    allowed: assetType.allowedSizes
                )
            }
            throw CombinerError.invalidMainTextureSize(name: name, size: size)
        }
    }

    static func validateInputSizes(
        _ inputs: [MapSlot: InputMap],
        targetSize: PixelSize,
        assetType: AssetType = .building
    ) throws {
        guard let baseColor = inputs[.baseColor] else {
            throw CombinerError.baseColorRequired
        }
        try validateMainInputSize(baseColor.size, name: baseColor.slot.title, assetType: assetType)
        guard targetSize == baseColor.size else {
            throw CombinerError.exportSizeMismatch(
                imported: baseColor.size,
                requested: targetSize
            )
        }

        let mismatches = inputs.values
            .filter { $0.size != baseColor.size }
            .sorted { $0.slot.title < $1.slot.title }
            .map { "\($0.slot.title): \($0.size)" }
        guard mismatches.isEmpty else {
            throw CombinerError.mismatchedTextureSizes(
                expected: baseColor.size,
                maps: mismatches
            )
        }
    }

    static func export(_ plan: TextureExportPlan) throws -> [URL] {
        try validateInputSizes(
            plan.inputs,
            targetSize: plan.targetSize,
            assetType: plan.assetType
        )

        let fileManager = FileManager.default
        try fileManager.createDirectory(
            at: plan.outputDirectory,
            withIntermediateDirectories: true
        )

        let staging = plan.outputDirectory.appendingPathComponent(
            ".cs2-combiner-\(UUID().uuidString)",
            isDirectory: true
        )
        try fileManager.createDirectory(at: staging, withIntermediateDirectories: false)
        defer { try? fileManager.removeItem(at: staging) }

        try writeBaseColor(
            plan,
            to: staging.appendingPathComponent(plan.outputName(suffix: "BaseColor"))
        )
        switch plan.assetType {
        case .building:
            try writeControlMask(
                plan,
                to: staging.appendingPathComponent(plan.outputName(suffix: "ControlMask"))
            )
            try writeBuildingMaskMap(
                plan,
                to: staging.appendingPathComponent(plan.outputName(suffix: "MaskMap"))
            )
        case .surface:
            try writeSurfaceMaskMap(
                plan,
                to: staging.appendingPathComponent(plan.outputName(suffix: "MaskMap"))
            )
        case .decal:
            if plan.outputSuffixes.contains("ControlMask") {
                try writeControlMask(
                    plan,
                    to: staging.appendingPathComponent(plan.outputName(suffix: "ControlMask"))
                )
            }
            try writeBuildingMaskMap(
                plan,
                to: staging.appendingPathComponent(plan.outputName(suffix: "MaskMap"))
            )
        }
        try writeNormal(
            plan,
            to: staging.appendingPathComponent(plan.outputName(suffix: "Normal"))
        )
        if plan.outputSuffixes.contains("Emissive") {
            try writeEmissive(
                plan,
                to: staging.appendingPathComponent(plan.outputName(suffix: "Emissive"))
            )
        }

        for name in plan.outputNames {
            let staged = staging.appendingPathComponent(name)
            let final = plan.outputDirectory.appendingPathComponent(name)
            if fileManager.fileExists(atPath: final.path) {
                try fileManager.removeItem(at: final)
            }
            try fileManager.moveItem(at: staged, to: final)
        }
        return plan.outputURLs
    }

    private static func writeBaseColor(_ plan: TextureExportPlan, to url: URL) throws {
        guard let baseInput = plan.inputs[.baseColor] else {
            throw CombinerError.baseColorRequired
        }
        let usesEmbeddedAlpha = try ImageLoader.hasAlphaChannel(baseInput.url)
        var base = try ImageLoader.raster(from: baseInput.url)
        let usesOpacityMap = plan.inputs[.opacity] != nil &&
            (!usesEmbeddedAlpha || plan.opacityMapOverridesBaseColorAlpha)
        let opacity = try usesOpacityMap ? plan.inputs[.opacity].map {
            try ImageLoader.raster(from: $0.url)
        } : nil

        if let opacity {
            for pixel in 0..<(plan.targetSize.width * plan.targetSize.height) {
                base.bytes[pixel * 4 + 3] = opacity.red(at: pixel)
            }
        } else if !usesEmbeddedAlpha {
            for pixel in 0..<(plan.targetSize.width * plan.targetSize.height) {
                base.bytes[pixel * 4 + 3] = 255
            }
        }
        try ImageLoader.writePNG(base, to: url)
    }

    private static func writeControlMask(_ plan: TextureExportPlan, to url: URL) throws {
        try ImageLoader.writePNG(controlMaskRaster(plan), to: url)
    }

    static func controlMaskRaster(_ plan: TextureExportPlan) throws -> ImageRaster {
        let channels = try [.cm1, .cm2, .cm3, .snowRemove].map { slot in
            try plan.inputs[slot].map {
                try ImageLoader.raster(from: $0.url)
            }
        }
        var output = blank(plan.targetSize, alpha: 0)
        for pixel in 0..<(plan.targetSize.width * plan.targetSize.height) {
            for channel in 0..<4 {
                output.bytes[pixel * 4 + channel] = channels[channel]?.red(at: pixel) ?? 0
            }
        }
        return output
    }

    private static func writeBuildingMaskMap(_ plan: TextureExportPlan, to url: URL) throws {
        let metallic = try raster(for: .metallic, in: plan)
        let coat = try raster(for: .coat, in: plan)
        let roughness = try raster(for: .roughness, in: plan)
        var output = blank(plan.targetSize, alpha: 0)
        for pixel in 0..<(plan.targetSize.width * plan.targetSize.height) {
            let index = pixel * 4
            output.bytes[index] = metallic?.red(at: pixel) ?? 0
            output.bytes[index + 1] = coat?.red(at: pixel) ?? 0
            output.bytes[index + 2] = 0
            output.bytes[index + 3] = 255 - (roughness?.red(at: pixel) ?? 255)
        }
        try ImageLoader.writePNG(output, to: url)
    }

    private static func writeSurfaceMaskMap(_ plan: TextureExportPlan, to url: URL) throws {
        try ImageLoader.writePNG(surfaceMaskRaster(plan), to: url)
    }

    static func surfaceMaskRaster(_ plan: TextureExportPlan) throws -> ImageRaster {
        let metallic = try raster(for: .metallic, in: plan)
        let metallicMask = try raster(for: .metallicMask, in: plan)
        let normalMask = try raster(for: .normalMask, in: plan)
        let roughness = try raster(for: .roughness, in: plan)
        var output = blank(plan.targetSize, alpha: 0)
        for pixel in 0..<(plan.targetSize.width * plan.targetSize.height) {
            let index = pixel * 4
            output.bytes[index] = metallic?.red(at: pixel) ?? 0
            output.bytes[index + 1] = metallicMask?.red(at: pixel) ?? 255
            output.bytes[index + 2] = normalMask?.red(at: pixel) ?? 255
            output.bytes[index + 3] = 255 - (roughness?.red(at: pixel) ?? 255)
        }
        return output
    }

    private static func writeNormal(_ plan: TextureExportPlan, to url: URL) throws {
        if var normal = try raster(for: .normal, in: plan) {
            if plan.normalizeNormalOnExport {
                NormalMapNormalization.normalize(&normal)
            }
            for pixel in 0..<(plan.targetSize.width * plan.targetSize.height) {
                normal.bytes[pixel * 4 + 3] = 255
            }
            try ImageLoader.writePNG(normal, to: url)
        } else {
            try ImageLoader.writePNG(
                solid(plan.targetSize, red: 128, green: 128, blue: 255),
                to: url
            )
        }
    }

    private static func writeEmissive(_ plan: TextureExportPlan, to url: URL) throws {
        if var emissive = try raster(for: .emissive, in: plan) {
            for pixel in 0..<(plan.targetSize.width * plan.targetSize.height) {
                emissive.bytes[pixel * 4 + 3] = 255
            }
            try ImageLoader.writePNG(emissive, to: url)
        } else {
            try ImageLoader.writePNG(solid(plan.targetSize, red: 0, green: 0, blue: 0), to: url)
        }
    }

    private static func raster(for slot: MapSlot, in plan: TextureExportPlan) throws -> ImageRaster? {
        try plan.inputs[slot].map {
            try ImageLoader.raster(from: $0.url)
        }
    }

    private static func blank(_ size: PixelSize, alpha: UInt8) -> ImageRaster {
        var raster = ImageRaster(
            width: size.width,
            height: size.height,
            bytes: [UInt8](repeating: 0, count: size.width * size.height * 4)
        )
        if alpha != 0 {
            for pixel in 0..<(size.width * size.height) {
                raster.bytes[pixel * 4 + 3] = alpha
            }
        }
        return raster
    }

    private static func solid(
        _ size: PixelSize,
        red: UInt8,
        green: UInt8,
        blue: UInt8
    ) -> ImageRaster {
        var raster = blank(size, alpha: 255)
        for pixel in 0..<(size.width * size.height) {
            let index = pixel * 4
            raster.bytes[index] = red
            raster.bytes[index + 1] = green
            raster.bytes[index + 2] = blue
        }
        return raster
    }
}

File: Windows/CS2Combiner.Core/TexturePacking.cs
namespace CS2Combiner.Core;

public static class TexturePacking
{
    public static PixelSize ValidateBaseColor(
        InputMap? input,
        AssetType assetType = AssetType.Building)
    {
        if (input is null)
        {
            throw new CombinerException("Add a BaseColor map before exporting.");
        }

        ValidateMainInputSize(input.Size, input.Slot.Title(), assetType);
        return input.Size;
    }

    public static void ValidateMainInputSize(
        PixelSize size,
        string name,
        AssetType assetType = AssetType.Building)
    {
        var allowedSizes = AssetProfiles.For(assetType).AllowedSizes;
        if (!size.IsSquare || !allowedSizes.Contains(size.Width))
        {
            if (assetType != AssetType.Building)
            {
                throw new CombinerException(
                    $"{name} for {assetType} must be square and exactly {string.Join(", ", allowedSizes)} pixels. It is {size}.");
            }
            throw new CombinerException(
                $"{name} must be square and exactly 512, 1024, 2048, or 4096 pixels. It is {size}.");
        }
    }

    public static void ValidateInputSizes(
        IReadOnlyDictionary<MapSlot, InputMap> inputs,
        PixelSize targetSize,
        AssetType assetType = AssetType.Building)
    {
        if (!inputs.TryGetValue(MapSlot.BaseColor, out var baseColor))
        {
            throw new CombinerException("Add a BaseColor map before exporting.");
        }

        ValidateMainInputSize(baseColor.Size, baseColor.Slot.Title(), assetType);

        if (targetSize != baseColor.Size)
        {
            throw new CombinerException(
                $"The export size {targetSize} does not match the imported BaseColor size {baseColor.Size}. Textures are never resized.");
        }

        var mismatches = inputs.Values
            .Where(input => input.Size != baseColor.Size)
            .OrderBy(input => input.Slot.Title(), StringComparer.CurrentCulture)
            .Select(input => $"{input.Slot.Title()}: {input.Size}")
            .ToArray();
        if (mismatches.Length > 0)
        {
            throw new CombinerException(
                $"All assigned main maps must match the BaseColor size {baseColor.Size}. Textures are never resized.\n\n" +
                string.Join('\n', mismatches));
        }
    }

    public static IReadOnlyList<string> Export(TextureExportPlan plan)
    {
        ValidateInputSizes(plan.Inputs, plan.TargetSize, plan.Profile);
        Directory.CreateDirectory(plan.OutputDirectory);
        var staging = Path.Combine(plan.OutputDirectory, $".cs2-combiner-{Guid.NewGuid():N}");
        Directory.CreateDirectory(staging);

        try
        {
            WriteBaseColor(plan, Path.Combine(staging, plan.OutputName("BaseColor")));
            switch (plan.Profile)
            {
                case AssetType.Building:
                    WriteControlMask(plan, Path.Combine(staging, plan.OutputName("ControlMask")));
                    WriteBuildingMaskMap(plan, Path.Combine(staging, plan.OutputName("MaskMap")));
                    break;
                case AssetType.Surface:
                    WriteSurfaceMaskMap(plan, Path.Combine(staging, plan.OutputName("MaskMap")));
                    break;
                case AssetType.Decal:
                    if (plan.ActiveOutputSuffixes.Contains("ControlMask"))
                    {
                        WriteControlMask(plan, Path.Combine(staging, plan.OutputName("ControlMask")));
                    }
                    WriteBuildingMaskMap(plan, Path.Combine(staging, plan.OutputName("MaskMap")));
                    break;
                default:
                    throw new ArgumentOutOfRangeException(nameof(plan.Profile), plan.Profile, null);
            }
            WriteNormal(plan, Path.Combine(staging, plan.OutputName("Normal")));
            if (plan.ActiveOutputSuffixes.Contains("Emissive"))
            {
                WriteEmissive(plan, Path.Combine(staging, plan.OutputName("Emissive")));
            }

            foreach (var name in plan.OutputNames)
            {
                File.Move(Path.Combine(staging, name), Path.Combine(plan.OutputDirectory, name), true);
            }
        }
        finally
        {
            if (Directory.Exists(staging))
            {
                Directory.Delete(staging, true);
            }
        }

        return plan.OutputPaths;
    }

    private static void WriteBaseColor(TextureExportPlan plan, string path)
    {
        if (!plan.Inputs.TryGetValue(MapSlot.BaseColor, out var baseInput))
        {
            throw new CombinerException("Add a BaseColor map before exporting.");
        }

        var usesEmbeddedAlpha = ImageCodec.HasAlphaChannel(baseInput.Path);
        var raster = ImageCodec.Load(baseInput.Path);
        var usesOpacityMap =
            plan.Inputs.ContainsKey(MapSlot.Opacity) &&
            (!usesEmbeddedAlpha || plan.OpacityMapOverridesBaseColorAlpha);
        var opacity = usesOpacityMap
            ? ImageCodec.Load(plan.Inputs[MapSlot.Opacity].Path)
            : null;

        for (var pixel = 0; pixel < plan.TargetSize.Width * plan.TargetSize.Height; pixel++)
        {
            if (opacity is not null)
            {
                raster.Bytes[pixel * 4 + 3] = opacity.Red(pixel);
            }
            else if (!usesEmbeddedAlpha)
            {
                raster.Bytes[pixel * 4 + 3] = byte.MaxValue;
            }
        }

        ImageCodec.WritePng(raster, path);
    }

    private static void WriteControlMask(TextureExportPlan plan, string path)
    {
        var channels = new[]
        {
            Raster(plan, MapSlot.ColorMask1),
            Raster(plan, MapSlot.ColorMask2),
            Raster(plan, MapSlot.ColorMask3),
            Raster(plan, MapSlot.SnowRemove)
        };
        var output = ImageRaster.Solid(plan.TargetSize, 0, 0, 0, 0);
        for (var pixel = 0; pixel < plan.TargetSize.Width * plan.TargetSize.Height; pixel++)
        {
            for (var channel = 0; channel < 4; channel++)
            {
                output.Bytes[pixel * 4 + channel] = channels[channel]?.Red(pixel) ?? 0;
            }
        }

        ImageCodec.WritePng(output, path);
    }

    private static void WriteBuildingMaskMap(TextureExportPlan plan, string path)
    {
        var metallic = Raster(plan, MapSlot.Metallic);
        var coat = Raster(plan, MapSlot.Coat);
        var roughness = Raster(plan, MapSlot.Roughness);
        var output = ImageRaster.Solid(plan.TargetSize, 0, 0, 0, 0);
        for (var pixel = 0; pixel < plan.TargetSize.Width * plan.TargetSize.Height; pixel++)
        {
            var index = pixel * 4;
            output.Bytes[index] = metallic?.Red(pixel) ?? 0;
            output.Bytes[index + 1] = coat?.Red(pixel) ?? 0;
            output.Bytes[index + 2] = 0;
            output.Bytes[index + 3] = (byte)(255 - (roughness?.Red(pixel) ?? 255));
        }

        ImageCodec.WritePng(output, path);
    }

    private static void WriteSurfaceMaskMap(TextureExportPlan plan, string path)
    {
        var metallic = Raster(plan, MapSlot.Metallic);
        var metallicMask = Raster(plan, MapSlot.MetallicMask);
        var normalMask = Raster(plan, MapSlot.NormalMask);
        var roughness = Raster(plan, MapSlot.Roughness);
        var output = ImageRaster.Solid(plan.TargetSize, 0, 0, 0, 0);
        for (var pixel = 0; pixel < plan.TargetSize.Width * plan.TargetSize.Height; pixel++)
        {
            var index = pixel * 4;
            output.Bytes[index] = metallic?.Red(pixel) ?? 0;
            output.Bytes[index + 1] = metallicMask?.Red(pixel) ?? 255;
            output.Bytes[index + 2] = normalMask?.Red(pixel) ?? 255;
            output.Bytes[index + 3] = (byte)(255 - (roughness?.Red(pixel) ?? 255));
        }

        ImageCodec.WritePng(output, path);
    }

    private static void WriteNormal(TextureExportPlan plan, string path)
    {
        var normal = Raster(plan, MapSlot.Normal);
        if (normal is null)
        {
            ImageCodec.WritePng(ImageRaster.Solid(plan.TargetSize, 128, 128, 255), path);
            return;
        }

        if (plan.NormalizeNormalOnExport)
        {
            NormalMapNormalization.Normalize(normal);
        }

        for (var pixel = 0; pixel < plan.TargetSize.Width * plan.TargetSize.Height; pixel++)
        {
            normal.Bytes[pixel * 4 + 3] = 255;
        }

        ImageCodec.WritePng(normal, path);
    }

    private static void WriteEmissive(TextureExportPlan plan, string path)
    {
        var emissive = Raster(plan, MapSlot.Emissive);
        if (emissive is null)
        {
            ImageCodec.WritePng(ImageRaster.Solid(plan.TargetSize, 0, 0, 0), path);
            return;
        }

        for (var pixel = 0; pixel < plan.TargetSize.Width * plan.TargetSize.Height; pixel++)
        {
            emissive.Bytes[pixel * 4 + 3] = 255;
        }

        ImageCodec.WritePng(emissive, path);
    }

    private static ImageRaster? Raster(TextureExportPlan plan, MapSlot slot) =>
        plan.Inputs.TryGetValue(slot, out var input) ? ImageCodec.Load(input.Path) : null;
}
